### PR TITLE
Auto-enable channels-last CUDA training

### PR DESCRIPTION
**Pull request by glenn-jocher**
## Summary

Make the existing tri-state `channels_last=None` default automatically select channels-last for CUDA training on PyTorch 1.11 and newer.

- `None`: automatic NHWC on CUDA with PyTorch 1.11+, unchanged elsewhere
- `False`: explicit NCHW opt-out
- `True`: preserves the existing explicit request behavior

The selection stays at the trainer setup owner and reuses the already-imported `TORCH_1_11` boundary. No new argument, helper, persistent state, or GPU allowlist is added.

## Compatibility envelope

RunPod validation covered the exact Platform training images and representative Ampere, Ada, Hopper, Blackwell, B200, and B300 GPUs across CUDA 11.1–13.2, PyTorch 1.8–2.12, and Python 3.8–3.13.

PyTorch 1.9 and 1.10 both reproduce a channels-last CUDA failure in YOLO26's detached one-to-one branch (`set_sizes_and_strides is not allowed on a Tensor created from .data or .detach()`). PyTorch 1.11 is the first passing version, so automatic selection begins there. On the proposed head, PyTorch 1.10 automatically remains NCHW and trains successfully, while PyTorch 1.11 automatically selects NHWC.

The exact proposed head passed:

- detect, segment, semantic, depth, classify, pose, and OBB training with automatic selection
- FP16 AMP, BF16, FP32, and `torch.compile`
- built-in and Ray tuner argument propagation (`None` remains automatic; explicit `False` is preserved)\n- two-GPU DDP
- checkpoint save and resume: checkpoints remain contiguous, resume reapplies automatic NHWC, and explicit `False` resumes in NCHW
- explicit `channels_last=False` opt-out

The production Platform image combinations passed at PyTorch 2.4/CUDA 12.4, PyTorch 2.8/CUDA 12.8, and PyTorch 2.12/CUDA 13.2.

## Performance

Paired YOLO26n/s/m loss + backward + optimizer-step timing at GPU-appropriate batches showed these channels-last speedup ranges:

| GPU architecture | Speedup |
| --- | ---: |
| Ampere | 1.17–1.25x |
| Ada | 1.15–1.18x |
| Hopper | 1.25–1.85x |
| RTX PRO Blackwell | 1.30–1.66x |
| B200 | 1.75–3.29x |
| B300 | 2.08–3.33x |

PyTorch 1.11 and 1.12 end-to-end paired training on RTX 3090 improved 1.26x and 1.23x respectively.

## Validation

- `pytest tests/test_python.py::test_autobackend_memory_format -q` — 1 passed
- CPU tri-state train smoke — `None`/`False` remain NCHW; explicit `True` warns and remains NCHW
- `tests/test_engine.py` — 32 passed; the sole depth label-histogram failure reproduces unchanged on `main` with the local NumPy/Matplotlib stack
- Ruff formatting, pinned Prettier 3.8.5, and `git diff --check`

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary

Training now automatically uses channels-last memory format for CUDA devices with PyTorch 1.11 or newer when `channels_last=None`, while preserving explicit opt-out behavior.

### 📊 Key Changes

- Updated `ultralytics/engine/trainer.py` to enable `torch.channels_last` when `channels_last=True`, or when it is `None` and `TORCH_1_11` is available, but only on CUDA.
- Preserved `channels_last=False` as an explicit NCHW opt-out and retained warnings when `channels_last=True` is requested on unsupported devices.
- Added `channels_last` to the resume override options so checkpoint resumes can reapply automatic selection or an explicit setting.
- Updated training documentation, argument descriptions, the default configuration comment, and the package version to `8.4.137`.

### 🎯 Purpose & Impact

- CUDA training on PyTorch 1.11 and newer now defaults to NHWC; PyTorch 1.10 and older, CPU, and MPS training remain NCHW by default.
- Users who need NCHW can set `channels_last=False`, while `channels_last=True` continues to explicitly request channels-last on CUDA.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @glenn-jocher, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This automated message confirms your contribution was received, and an Ultralytics engineer will assist with the review. To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Reviewed all four changed files and the surrounding trainer, configuration, validation, checkpoint, and resume paths. The automatic CUDA selection is correct for new training runs, but the explicit `channels_last=False` opt-out is not honored when resuming because resume argument restoration drops that override.

<details><summary>💬 Posted 1 inline comment</summary>

- 💡 **MEDIUM** `ultralytics/engine/trainer.py:321` On resume, `check_resume()` replaces `self.args` with the checkpoint arguments and its override allowlist omits `channels_last`. Therefore `yolo train resume model=last.pt channels_last=False` is silently reset to the checkpoint's `None`; on CUDA with PyTorch 1.11+ this expression enables NHWC, so the documented explicit NCHW opt-out fails for resumed runs. Preserve `channels_last` in the resume override handling.
</details>

File: ultralytics/engine/trainer.py [hunk 1]
```diff
@@ -318,7 +318,8 @@ def _setup_train(self):
         self.model = self.model.to(self.device)
         # channels_last (NHWC) is CUDA-only: lossless and Tensor-Core friendly there, but numerically wrong
         # on MPS and no benefit on CPU
-        if self.args.channels_last and self.device.type == "cuda":
+        channels_last = self.args.channels_last is True or (self.args.channels_last is None and TORCH_1_11)
```
Comment (UltralyticsAssistant) [inline ultralytics/engine/trainer.py:321]: 💡 **MEDIUM**: On resume, `check_resume()` replaces `self.args` with the checkpoint arguments and its override allowlist omits `channels_last`. Therefore `yolo train resume model=last.pt channels_last=False` is silently reset to the checkpoint's `None`; on CUDA with PyTorch 1.11+ this expression enables NHWC, so the documented explicit NCHW opt-out fails for resumed runs. Preserve `channels_last` in the resume override handling.

Comment (glenn-jocher) [reply ultralytics/engine/trainer.py:321]: Fixed in 311a6a0d5. `channels_last` now uses the existing resume override owner. Verified on the exact head with the Platform Ada image (PyTorch 2.4.1+cu124, Python 3.11, RTX 4090): an auto-NHWC checkpoint remains contiguous on disk and `resume=True channels_last=False` resumes with NCHW weights.

**Comment by glenn-jocher** [comment]
## ⚡ Actions Trigger

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

GitHub Actions below triggered via workflow dispatch for this PR at 2026-08-31 16:32:19 UTC with `@ultralytics/run-all` command
(available commands are `@ultralytics/run-all`, `@ultralytics/run-ci`, and `@ultralytics/run-docker`):

* ✅ [CI](https://github.com/ultralytics/ultralytics/actions/runs/33414556162): `ci.yml` run 30911
* ✅ [Publish Docker Images](https://github.com/ultralytics/ultralytics/actions/runs/33414558150): `docker.yml` run 3132

**Comment by UltralyticsAssistant** [comment]
🎉 This PR has been successfully merged—thank you, @glenn-jocher!

> “Alone we can do so little; together we can do so much.” — Helen Keller

This collaboration makes CUDA training automatically use channels-last memory format with PyTorch 1.11+ while preserving explicit NCHW opt-out behavior and consistent resume settings. See [PR #26007](https://github.com/ultralytics/ultralytics/pull/26007).
